x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindFieldStructureAction.java

```diff
@@ -0,0 +1,105 @@
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.ml.utils.MapHelper;
+import org.elasticsearch.xpack.core.textstructure.action.AbstractFindStructureRequest;
+import org.elasticsearch.xpack.core.textstructure.action.FindFieldStructureAction;
+import org.elasticsearch.xpack.core.textstructure.action.FindStructureResponse;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureFinder;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureFinderManager;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureOverrides;
+import org.elasticsearch.xpack.textstructure.structurefinder.TimeoutChecker;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.threadpool.ThreadPool.Names.GENERIC;
+
+public class TransportFindFieldStructureAction extends HandledTransportAction<FindFieldStructureAction.Request, FindStructureResponse> {
+
+    private final Client client;
+    private final ThreadPool threadPool;
+
+    @Inject
+    public TransportFindFieldStructureAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        Client client,
+        ThreadPool threadPool
+    ) {
+        super(
+            FindFieldStructureAction.NAME,
+            transportService,
+            actionFilters,
+            FindFieldStructureAction.Request::new,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.client = client;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    protected void doExecute(Task task, FindFieldStructureAction.Request request, ActionListener<FindStructureResponse> listener) {
+        client.prepareSearch(request.getIndex())
+            .setSize(request.getLinesToSample())
+            .setFetchSource(true)
+            .setQuery(QueryBuilders.existsQuery(request.getField()))
+            .setFetchSource(new String[] { request.getField() }, null)
+            .execute(ActionListener.wrap(searchResponse -> {
+                long hitCount = searchResponse.getHits().getHits().length;
+                if (hitCount < AbstractFindStructureRequest.MIN_SAMPLE_LINE_COUNT) {
+                    throw new IllegalArgumentException("Input contained too few lines [" + hitCount + "] to obtain a meaningful sample");
```

Comment:
It's inconsistent to throw here when a couple of lines below we explicitly catch and pass on to the listener.

```suggestion
                    listener.onFailure(new IllegalArgumentException("Input contained too few lines [" + hitCount + "] to obtain a meaningful sample"));
                    return;
```

Both ways do the [same thing](https://github.com/elastic/elasticsearch/blob/cf05e28f9c337edd7e10518fa26dd8eec08a2098/server/src/main/java/org/elasticsearch/action/ActionListener.java#L169), but it's more traditional to pass exceptions directly to listeners rather than throw from methods that know which listener they should reply to. (I guess it's slightly faster to avoid the throw/catch overhead.)